tomb-common/src/metadata.rs

```diff
@@ -423,149 +399,216 @@ impl FsMetadata {
+            )
+            .await
     }
 
     /// Rm a file or directory
     pub async fn rm(
         &mut self,
-        path_segments: Vec<String>,
+        path_segments: &[String],
         store: &impl RootedBlockStore,
     ) -> Result<()> {
-        // Search through the PrivateDirectory for a Node that matches the path provided
-        let result = self
-            .root_dir
-            .get_node(&path_segments, true, &self.metadata_forest, store)
-            .await;
-
-        if let Ok(node) = result && node.is_some() {
-            // Create the subdirectory
-            self.root_dir
-                .rm(
-                    &path_segments,
-                    true,
-                    &self.metadata_forest,
-                    store,
-                )
-                .await?;
-        } else {
-            return Err(SerialError::NodeNotFound(
-                path_segments.join("/"),
-            )
-            .into());
-        }
-        Ok(())
+        // Create the subdirectory
+        self.root_dir
+            .rm(path_segments, true, &self.forest, store)
+            .await
+            .map(|_| ())
+            .map_err(|_| SerialError::NodeNotFound(path_segments.join("/")).into())
     }
 
     /// Add a Vector of bytes as a new file in the Fs. Store in our content store
     pub async fn read(
         &mut self,
-        path_segments: Vec<String>,
+        path_segments: &[String],
         metadata_store: &impl RootedBlockStore,
         content_store: &impl BlockStore,
     ) -> Result<Vec<u8>> {
         // Compress the data in the file
         let result = self
             .root_dir
-            .get_node(&path_segments, true, &self.metadata_forest, metadata_store)
+            .get_node(path_segments, true, &self.forest, metadata_store)
             .await
             .expect("node not found");
-        match result {
-            Some(PrivateNode::File(file)) => {
-                let content = file
-                    .get_content(&self.content_forest, content_store)
-                    .await?;
-                Ok(content)
-            }
-            _ => Err(SerialError::NodeNotFound(path_segments.join("/")).into()),
+
+        // Split store
+        let split_store = DoubleSplitStore::new(metadata_store, content_store);
+
+        // If the node is found and is a file
+        if let Some(PrivateNode::File(file)) = result {
+            let content = file.get_content(&self.forest, &split_store).await?;
+            Ok(content)
+        } else {
+            Err(SerialError::NodeNotFound(path_segments.join("/")).into())
+        }
+    }
+
+    /// Write data do a specific node
+    pub async fn write(
+        &mut self,
+        path_segments: &[String],
+        metadata_store: &impl RootedBlockStore,
+        content_store: &impl BlockStore,
+        content: Vec<u8>,
+    ) -> Result<()> {
+        let time = Utc::now();
+        let rng = &mut thread_rng();
```

Comment:
This is the same pattern of taking a reference to an unowned object, this one is mutable so passing it around always ensure the caller gets a mutable version even if they don't need it. It become impossible to tell where mutations occur.